feat(ui): Plugin support for menu items

milesap opened the pull request:
## Description
Added working demo to index-fgp-en.html with back to cart functionality. 

Closes #1438

## Testing
:eyes: 

## Documentation
Will document the plugin classes and create a wiki page once initial reviews are complete.

## Checklist
<!-- Quick checklist for items that are easy to miss -->

- [x] commits messages follow the guidelines
- [x] code passes unit tests
- [x] release notes have been updated
- [x] PR targets the correct release version

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/1490)
<!-- Reviewable:end -->

james-rae commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 1 of 7 files at r1, 8 of 10 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[gulpfile.js, line 306 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3afuF933ZOfn6S8sd:-KZ3afuF933ZOfn6S8se:bem2nds) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/gulpfile.js#L306)):*
> ```JavaScript
>         // if it gets bigger we should split it into a separate file
>         const injector = merge(
>             gulp.src(config.jsPluginSupport),
> ```

this should probably go through babel as well

---

*[src/index-fgp-en.html, line 558 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bEEf3WttdkobOR9a:-KZ3bEEgAo6HexmFf4vz:b-1bq2b) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/index-fgp-en.html#L558)):*
> ```HTML
>         const BackToCartPlugin = new RV.Plugin.MenuItem(
>             'cartButtonLabel', // the translation key or a string label
>             () => RV.getMap('fgpmap').backToCart().then(keys => window.location.href = window.location.href.split('?')[0] + '?keys=' + keys.toString()),
> ```

Might be missing something, but isn't part of the goal to remove backToCart from the external API?

---

*[src/app/bootstrap.js, line 130 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bv4y4yke-74xDy_5:-KZ3bv4y4yke-74xDy_6:b-ufknca) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/bootstrap.js#L130)):*
> ```JavaScript
>         },
> 
>         registerPlugin(plugin) {
> ```

I think registerPlugin should take a unique identifier as the plugin name (we can add a prefix ourselves to avoid conflicts, but I think it will be easier than trying to hunt down `plugin-2223` when things start going wrong).

---

*[src/app/bootstrap.js, line 133 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bYpnsWtpea1CoLJh:-KZ3bYpoxu-hGvUsM56Q:b-rf6kk2) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/bootstrap.js#L133)):*
> ```JavaScript
>             this._initProxy('translationService', plugin.translations).then(() => {
>                 RV.plugins[this.appID].push(plugin);
>                 if (RV.evtBindings.onCreate[plugin.constructor.name][this.appID]) {
> ```

evtBindings should not be exposed through `RV`

---

*[src/app/plugin.js, line 20 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3eTwEO-ysakM3gKyB:-KZ3eTwEO-ysakM3gKyC:b-oc1wjg) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/plugin.js#L20)):*
> ```JavaScript
>         }
> 
>         translationKey (key) {
> ```

Will this be consumed by anything other than the subclasses?

---

*[src/app/plugin.js, line 36 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3es1G5KGeT8HRCCAv:-KZ3es1G5KGeT8HRCCAw:b-jj008h) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/plugin.js#L36)):*
> ```JavaScript
>         constructor (name, actionCB, translations) {
>             super(translations);
>             this.name = this.translationKey(name);
> ```

Can you convert these to ES6 properties?

---

*[src/app/plugin.js, line 47 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3f20rsJngahZzpKeX:-KZ3f20rsJngahZzpKeY:bzhp68e) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/plugin.js#L47)):*
> ```JavaScript
> 
>     RV.Plugin = {
>         BasePlugin,
> ```

I don't think BasePlugin should be exposed

---

*[src/app/core/translation.service.js, line 4 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3d7EKHcmNKzTaoqCS:-KZ3d7EKHcmNKzTaoqCT:b-sl9ipc) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/core/translation.service.js#L4)):*
> ```JavaScript
> 
>     /**
>      * @ngdoc service
> ```

This follows the dgeni doc guidelines, needs to be switched to jsdoc

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

milesap commented:
Review status: 3 of 9 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[gulpfile.js, line 306 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3afuF933ZOfn6S8sd:-KZ72TpHFaDctGW80j-I:b-hmjed7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/gulpfile.js#L306)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

this should probably go through babel as well
</blockquote></details>

Added babel

---

*[src/index-fgp-en.html, line 558 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bEEf3WttdkobOR9a:-KZ72hUMoanssWsRrztW:bt2rpjh) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/index-fgp-en.html#L558)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Might be missing something, but isn't part of the goal to remove backToCart from the external API?
</blockquote></details>

Well back to cart and it's functionality should be a plugin, so it would be outside the scope of the viewers API. At least that's my take on it, unless I'm missing something as well :smile: 

---

*[src/app/bootstrap.js, line 130 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bv4y4yke-74xDy_5:-KZ78QfoxXY4PWbyIAGg:b-r06c5l) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/bootstrap.js#L130)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

I think registerPlugin should take a unique identifier as the plugin name (we can add a prefix ourselves to avoid conflicts, but I think it will be easier than trying to hunt down `plugin-2223` when things start going wrong).
</blockquote></details>

Good idea - added plugin name to `registerPlugin` 

---

*[src/app/bootstrap.js, line 133 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bYpnsWtpea1CoLJh:-KZ78HwbuHX2IFtT8K8Q:b-ljgox3) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/bootstrap.js#L133)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

evtBindings should not be exposed through `RV`
</blockquote></details>

Removed from RV, now accessible through static method `eventBindings`

---

*[src/app/plugin.js, line 20 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3eTwEO-ysakM3gKyB:-KZ7CMa7kFek11J8v8f6:b-8dnznt) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/plugin.js#L20)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Will this be consumed by anything other than the subclasses?
</blockquote></details>

This has been replaced by `translateProperty` since the full plugin ID is not known until `registerPlugin` is called with the plugin name.

---

*[src/app/plugin.js, line 36 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3es1G5KGeT8HRCCAv:-KZ7ChLBGpFNxJ6CQ9WU:b42o1du) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/plugin.js#L36)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Can you convert these to ES6 properties?
</blockquote></details>

Not sure what you mean here, maybe an example would help

---

*[src/app/plugin.js, line 47 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3f20rsJngahZzpKeX:-KZ7CtEo6-CbzH1liW5E:bu013vn) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/plugin.js#L47)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

I don't think BasePlugin should be exposed
</blockquote></details>

No longer exposed

---

*[src/app/core/translation.service.js, line 4 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3d7EKHcmNKzTaoqCS:-KZ7DR417s83FKwvHlH4:bv51zn2) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/core/translation.service.js#L4)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

This follows the dgeni doc guidelines, needs to be switched to jsdoc
</blockquote></details>

Ah, ok corrected the doc

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 3 of 6 files at r3.
Review status: 6 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/index-fgp-en.html, line 558 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bEEf3WttdkobOR9a:-KZ7EWxmznOlXqjjmYQC:b-lfgse7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/index-fgp-en.html#L558)):*
> RV.getMap('fgpmap').backToCart()

This the the part I'm questioning.  This plugin seems like it's wrapping the functionality already in the external API instead of refactoring the functionality into itself.

---

*[src/app/bootstrap.js, line 130 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bv4y4yke-74xDy_5:-KZ7FfdTwbB-AhKMTFVP:b6hf5ws) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/bootstrap.js#L130)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Good idea - added plugin name to `registerPlugin` 
</blockquote></details>

If `pluginName` is unique (and I think we should enforce this) then `pluginNum` is redundant.

---

*[src/app/plugin.js, line 20 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3eTwEO-ysakM3gKyB:-KZ7JIr-N3IfQeDeRfjO:b4hslez) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/plugin.js#L20)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

This has been replaced by `translateProperty` since the full plugin ID is not known until `registerPlugin` is called with the plugin name.
</blockquote></details>

Based on our discussion about IDs I think this can also become an ES6 property.

---

*[src/app/plugin.js, line 36 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3es1G5KGeT8HRCCAv:-KZ7GIfHJ1nu1X1ADZuA:b-teizmu) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/plugin.js#L36)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Not sure what you mean here, maybe an example would help
</blockquote></details>

https://github.com/fgpv-vpgf/fgpv-vpgf/blob/v1.4.0/src/app/geo/layer-record.class.js#L18

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

milesap commented:
For `evtBindings` - we wanted to move it into `bootstrap.js`. However `evtBindings` is needed in `plugin.js` for the static `onCreate` method, so I wasn't quite sure what you had in mind for this change.

---

Review status: 5 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/index-fgp-en.html, line 558 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bEEf3WttdkobOR9a:-KZMaBKgxG_6sQCCroWR:b-75nnr4) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/index-fgp-en.html#L558)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

> RV.getMap('fgpmap').backToCart()

This the the part I'm questioning.  This plugin seems like it's wrapping the functionality already in the external API instead of refactoring the functionality into itself.
</blockquote></details>

We can discuss this briefly in the weekly code review - extracting `backToCart` from the viewer to the plugin will require access to `bookmarkService` and `geoService` from the plugin.

---

*[src/app/bootstrap.js, line 130 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bv4y4yke-74xDy_5:-KZMbKWpBsPzaHz3UAQ1:b-f7sawo) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/bootstrap.js#L130)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

If `pluginName` is unique (and I think we should enforce this) then `pluginNum` is redundant.
</blockquote></details>

Removed `pluginNum`

---

*[src/app/plugin.js, line 20 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3eTwEO-ysakM3gKyB:-KZMbPivorfFh_oMZ-dV:bpvwwec) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/plugin.js#L20)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Based on our discussion about IDs I think this can also become an ES6 property.
</blockquote></details>

added a few ES6 properties, let me know if I missed any

---

*[src/app/plugin.js, line 36 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3es1G5KGeT8HRCCAv:-KZMbUYR6oYPQ0F5kx5l:bm0ddos) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/app/plugin.js#L36)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

https://github.com/fgpv-vpgf/fgpv-vpgf/blob/v1.4.0/src/app/geo/layer-record.class.js#L18
</blockquote></details>

ty this helped

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-:-KZMbZ_Lj3OwlvSWfIBG:bx47jy0)*
<!-- Sent from Reviewable.io -->

james-rae commented:
Reviewed 3 of 10 files at r2, 3 of 6 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/app/bootstrap.js, line 135 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZRqn3FCv8_fHi5fyi2:-KZRqn3FCv8_fHi5fyi3:bqudqad) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/1bd2c6fde4bfd68a22e03a13ff1e52eb12734cc6/src/app/bootstrap.js#L135)):*
> ```JavaScript
>             }
> 
>             RV.plugins[this.appID].push(plugin);
> ```

Do we need to keep the plugins separately? Can we store them inside the viewer instance? This will simplify access, as there is no need to check what `appId` it belongs to.

---

*[src/app/bootstrap.js, line 137 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZRpfESeGmyAv9myhya:-KZRpfESeGmyAv9myhyb:b-kb2nl6) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/1bd2c6fde4bfd68a22e03a13ff1e52eb12734cc6/src/app/bootstrap.js#L137)):*
> ```JavaScript
>             RV.plugins[this.appID].push(plugin);
> 
>             plugin.id = pluginName;
> ```

If `pluginName` is plugin `id`, why not just pass it to the constructor?

If I try to add the same plugin to two different maps,
```js
map1.registerPlugin('pluginOne', plugin);
map2.registerPlugin('pluginTwo', plugin); 
```

I think it will actually override plugin's id with 'pluginTwo'.

---

*[src/app/layout/shell.directive.js, line 89 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZRoGawEVCQNqTPFIog:-KZRoGawEVCQNqTPFIoh:b-c55ca5) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/1bd2c6fde4bfd68a22e03a13ff1e52eb12734cc6/src/app/layout/shell.directive.js#L89)):*
> ```JavaScript
> 
>         // Add any MenuItem plugins as they are created to the menu
>         globalRegistry.Plugin.MenuItem.onCreate(appInfo.id, mItem => self.menu[0].children.push(mItem));
> ```

Maybe for the future - let the user specify position of the plugin menu item.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

milesap commented:
Note that `evtBindings` have been replaced with the implementation of `plugin.service.js`.

---

Review status: 4 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/app/bootstrap.js, line 135 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZRqn3FCv8_fHi5fyi2:-KZXl-gD6HtuyUSmcFdE:b-mek99y) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/1bd2c6fde4bfd68a22e03a13ff1e52eb12734cc6/src/app/bootstrap.js#L135)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary><blockquote>

Do we need to keep the plugins separately? Can we store them inside the viewer instance? This will simplify access, as there is no need to check what `appId` it belongs to.
</blockquote></details>

Plugins are now registered directly in `plugin.service.js` - which is proxied through `bootstrap.js`.

---

*[src/app/bootstrap.js, line 137 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZRpfESeGmyAv9myhya:-KZXlBK6mO3NuY7q0jZw:b-cdj55x) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/1bd2c6fde4bfd68a22e03a13ff1e52eb12734cc6/src/app/bootstrap.js#L137)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary><blockquote>

If `pluginName` is plugin `id`, why not just pass it to the constructor?

If I try to add the same plugin to two different maps,
```js
map1.registerPlugin('pluginOne', plugin);
map2.registerPlugin('pluginTwo', plugin); 
```

I think it will actually override plugin's id with 'pluginTwo'.
</blockquote></details>

Changed so that pluginID is now passed as the first parameter during object initialization. This should also resolve the issue you have shown above.

---

*[src/app/layout/shell.directive.js, line 89 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZRoGawEVCQNqTPFIog:-KZWKQkXluYeFsgFAPak:b2qd9vd) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/1bd2c6fde4bfd68a22e03a13ff1e52eb12734cc6/src/app/layout/shell.directive.js#L89)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary><blockquote>

Maybe for the future - let the user specify position of the plugin menu item.
</blockquote></details>

:+1: good idea

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-:-KZXkjddkSbIQcXkJ-ur:b-54yl0k)*
<!-- Sent from Reviewable.io -->

james-rae commented:
Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 1 of 3 files at r4, 5 of 6 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/app/plugin.js, line 43 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZMcxHib85TUfawiue_:-KZMcxHib85TUfawiuea:byf2g8c) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/1bd2c6fde4bfd68a22e03a13ff1e52eb12734cc6/src/app/plugin.js#L43)):*
> ```JavaScript
> 
>     class MenuItem extends BasePlugin {
>         get type () { return this._type; }
> ```

Can you move this into the base class (you can implement it throwing an exception since it should be overridden by every sub class).

---

*[src/app/plugin.js, line 3 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2QK5L5P2Ewnjl_mG6:-Ka2QK5L5P2Ewnjl_mG7:b-pxp1db) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/plugin.js#L3)):*
> ```JavaScript
> /* global RV */
> (() => {
>     class BasePlugin {
> ```

Can you add the type property here (it can throw an exception if called), but it makes it clear that everything under it should implement it.  Basically a weak version of an abstract property for JS.

Name can probably go here with the full code since I think all plugins will be named.

---

*[src/app/core/plugin.service.js, line 30 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2OaRJEKRDwZl9bRsK:-Ka2OaRJEKRDwZl9bRsL:b-dkfkfn) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/core/plugin.service.js#L30)):*
> ```JavaScript
>         function register(plugin) {
>             if (!(plugin instanceof RV.Plugin.BasePlugin)) {
>                 throw 'Plugin is not an extension of BasePlugin.';
> ```

Please wrap these in `throw new Error()` instead of throwing strings.

---

*[src/app/core/plugin.service.js, line 47 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2PsYLk3BWuLwKaFiy:-Ka2PsYLk3BWuLwKaFiz:bf5za85) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/core/plugin.service.js#L47)):*
> ```JavaScript
>             _onCreate[pType].push(cb);
>             // trigger this callback for any plugin already created
>             pluginList.forEach(pi => pi.constructor.name === pType ? cb(pi) : undefined);
> ```

Style suggestion: `list.filter(name = type).forEach(callback)` would get rid of the odd looking `: undefined`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Reviewed 2 of 10 files at r2, 1 of 6 files at r3, 1 of 3 files at r4, 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/app/layout/shell.directive.js, line 89 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2XFCiEZQhztByZW5-:-Ka2XFCjlDA78kyYyxOO:byhvcu8) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/layout/shell.directive.js#L89)):*
> ```JavaScript
> 
>         // Add any MenuItem plugins as they are created to the menu
>         pluginService.onCreate('MenuItem', mItem => self.menu[0].children.push(mItem));
> ```

This part will probably needs to be updated since the side menu refactor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

milesap commented:
Review status: 5 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/app/plugin.js, line 43 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZMcxHib85TUfawiue_:-Ka3WqjUWql5sm4qraSx:b4qvl1) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/1bd2c6fde4bfd68a22e03a13ff1e52eb12734cc6/src/app/plugin.js#L43)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Can you move this into the base class (you can implement it throwing an exception since it should be overridden by every sub class).
</blockquote></details>

Made the change

---

*[src/app/plugin.js, line 3 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2QK5L5P2Ewnjl_mG6:-Ka3YdSFKgD0fJH2eMM3:bla1y27) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/plugin.js#L3)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Can you add the type property here (it can throw an exception if called), but it makes it clear that everything under it should implement it.  Basically a weak version of an abstract property for JS.

Name can probably go here with the full code since I think all plugins will be named.
</blockquote></details>

`type` and `name` properties moved into `BasePlugin`

---

*[src/app/core/plugin.service.js, line 30 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2OaRJEKRDwZl9bRsK:-Ka3Wm7sFXQcloKbKPFL:bngg66u) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/core/plugin.service.js#L30)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Please wrap these in `throw new Error()` instead of throwing strings.
</blockquote></details>

Wrapped in `throw new Error()`

---

*[src/app/core/plugin.service.js, line 47 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2PsYLk3BWuLwKaFiy:-Ka3WhCy-dNoA3iuhMSq:b-i5vsya) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/core/plugin.service.js#L47)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Style suggestion: `list.filter(name = type).forEach(callback)` would get rid of the odd looking `: undefined`.
</blockquote></details>

Great, made the change

---

*[src/app/layout/shell.directive.js, line 89 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2XFCiEZQhztByZW5-:-Ka3kYTSjY5Ra5o-9U7-:b-eab352) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/layout/shell.directive.js#L89)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary><blockquote>

This part will probably needs to be updated since the side menu refactor.
</blockquote></details>

Plugin links go into the plugins group which appears at the bottom. I'll push another update tomorrow so that the plugin group is hidden if there are no plugins.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 7 of 8 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/app/plugin.js, line 43 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZMcxHib85TUfawiue_:-Ka3lqTTQMfO4nj0DDxd:b-1x671w) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/1bd2c6fde4bfd68a22e03a13ff1e52eb12734cc6/src/app/plugin.js#L43)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Made the change
</blockquote></details>

Quick follow up based on our pod discussion.  Is `type` being used at the moment (and if not maybe we should comment it until we need it)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

milesap commented:
Plugin group in sidenav will no longer display if there are no plugin `MenuItem` registered.

---

Review status: 11 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/app/plugin.js, line 43 at r4](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZMcxHib85TUfawiue_:-Ka7idXtx_nieKTek5r0:b-vj7cyj) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/1bd2c6fde4bfd68a22e03a13ff1e52eb12734cc6/src/app/plugin.js#L43)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Quick follow up based on our pod discussion.  Is `type` being used at the moment (and if not maybe we should comment it until we need it)?
</blockquote></details>

In this case `type` is specific to `MenuItem`. It's value can be either `group` or `link`. I removed it from `BasePlugin` but left it in `MenuItem`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-:-Ka7jHjN99AVDRrCWUWE:bzbczjp)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Review status: 11 of 13 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/app/layout/shell.directive.js, line 89 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2XFCiEZQhztByZW5-:-Ka7rxeHdeznGdi275DH:bv05w0p) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/layout/shell.directive.js#L89)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Plugin links go into the plugins group which appears at the bottom. I'll push another update tomorrow so that the plugin group is hidden if there are no plugins.
</blockquote></details>

Ideally, plugins group should only be created if there is a plugin added to it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

jolevesq commented:
Reviewed 1 of 10 files at r2, 1 of 6 files at r3, 2 of 3 files at r4, 2 of 6 files at r5, 6 of 8 files at r6, 1 of 2 files at r7.
Review status: 12 of 13 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Reviewed 6 of 8 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 1 of 6 files at r5, 1 of 8 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/index-fgp-en.html, line 558 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bEEf3WttdkobOR9a:-Ka7t139ri-eAb2OVDlN:b9k31t9) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/index-fgp-en.html#L558)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

We can discuss this briefly in the weekly code review - extracting `backToCart` from the viewer to the plugin will require access to `bookmarkService` and `geoService` from the plugin.
</blockquote></details>

Based on discussion I'm splitting this into a separate issue #1565 .  I'd like to add one extra bit along with this PR to deprecate the current API call.  Can you add a `console.warn` to https://github.com/fgpv-vpgf/fgpv-vpgf/blob/develop/src/app/bootstrap.js#L125 so anyone calling backToCart knows that it's going away?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

alyec commented:
I don't see the plugin on fgp-en demo and it looks like a bunch of code got removed from sidenav.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-:-Ka8DBBClkvt8_flpVJM:b-yc6w3a)*
<!-- Sent from Reviewable.io -->

milesap commented:
Review status: 10 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/index-fgp-en.html, line 558 at r2](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-KZ3bEEf3WttdkobOR9a:-Ka8cbe4qlEC_MMjzE0_:b-ggkp0t) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/9b577e1daf6d44a3dd5583455074c89033714c12/src/index-fgp-en.html#L558)):*
<details><summary><i>Previously, alyec (Aly Merchant) wrote…</i></summary><blockquote>

Based on discussion I'm splitting this into a separate issue #1565 .  I'd like to add one extra bit along with this PR to deprecate the current API call.  Can you add a `console.warn` to https://github.com/fgpv-vpgf/fgpv-vpgf/blob/develop/src/app/bootstrap.js#L125 so anyone calling backToCart knows that it's going away?
</blockquote></details>

Added a deprecated warning

---

*[src/app/layout/shell.directive.js, line 89 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2XFCiEZQhztByZW5-:-Ka8fApjPIGnieGml21r:blcndko) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/layout/shell.directive.js#L89)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary><blockquote>

Ideally, plugins group should only be created if there is a plugin added to it.
</blockquote></details>

Ok changed so that plugin group is only created when a plugin is added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Reviewed 2 of 3 files at r8.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/app/ui/sidenav/sidenav.service.js, line 138 at r8](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka8fuXz0wTew9R6ITze:-Ka8fuY-2O1CG7ynkIVr:bx979su) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/3fa4059978e88cf3d00bf97096dce44e122642a3/src/app/ui/sidenav/sidenav.service.js#L138)):*
> ```JavaScript
>                 label: 'sidenav.label.language',
>                 icon: 'action:translate',
>                 children: []
> ```

You can keep the plugins group definition here, it's just for reference and will only be used when you push `plugins` into `SIDENAV_CONFIG_DEFAULT.items`.

---

*[src/app/layout/shell.directive.js, line 89 at r5](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka2XFCiEZQhztByZW5-:-Ka8goNn6XzrpRfeijiG:b-8fx2u7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/0a7511a6202eae1e20307ecd6cfe03e9a6558bae/src/app/layout/shell.directive.js#L89)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Ok changed so that plugin group is only created when a plugin is added.
</blockquote></details>

:+1: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

milesap commented:
Review status: 11 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/app/ui/sidenav/sidenav.service.js, line 138 at r8](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka8fuXz0wTew9R6ITze:-Ka8i0ftcavOYot1XVUR:bijbbcs) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/3fa4059978e88cf3d00bf97096dce44e122642a3/src/app/ui/sidenav/sidenav.service.js#L138)):*
<details><summary><i>Previously, AleksueiR (Aleksuei Riabtsev) wrote…</i></summary><blockquote>

You can keep the plugins group definition here, it's just for reference and will only be used when you push `plugins` into `SIDENAV_CONFIG_DEFAULT.items`.
</blockquote></details>

Ok made change

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

milesap commented:
It works on my machine but not on demo links? 

---

Review status: 11 of 12 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-:-Ka8i7gvc_2JuxqDp1xP:by5jd7c)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/app/ui/sidenav/sidenav.service.js, line 138 at r8](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-Ka8fuXz0wTew9R6ITze:-Ka8itzHtzsQx6V0z5ut:b-8fx2u7) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/3fa4059978e88cf3d00bf97096dce44e122642a3/src/app/ui/sidenav/sidenav.service.js#L138)):*
<details><summary><i>Previously, milesap (Miles Petrov) wrote…</i></summary><blockquote>

Ok made change
</blockquote></details>

:+1: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

milesap commented:
:fire: Still no success on prod... :fire: 

---

Review status: 8 of 12 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-:-KaDqlrBEG0yHRFrg2B5:b-fsog8x)*
<!-- Sent from Reviewable.io -->

milesap commented:
:pray: 

---

Review status: 7 of 12 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-:-KaHm25AJ9g4KQ7_QYHK:bi9estx)*
<!-- Sent from Reviewable.io -->

alyec commented:
Reviewed 2 of 3 files at r8, 2 of 4 files at r10, 3 of 3 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

milesap commented:
The stars have aligned - dev and prod builds work

---

Review status: 11 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-:-KaJD2_nNbhhjP4G8D8q:b-eqrhxx)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
Reviewed 2 of 4 files at r10, 3 of 3 files at r11, 1 of 1 files at r12.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490)*
<!-- Sent from Reviewable.io -->

AleksueiR commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-:-KaJDPnlyqeTgKxqQ97q:bnfp4nl)*
<!-- Sent from Reviewable.io -->

alyec commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (checked dev builds)

---

Reviewed 1 of 1 files at r12.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fgpv-vpgf/fgpv-vpgf/1490#-:-KaN2uTVqQHhcho7JFOc:b-dij3t)*
<!-- Sent from Reviewable.io -->
